### ISSUE TITLE
Adding zib ValueSets to CodeableConcepts with required bindings

### ISSUE THREAD

**Issue by pieter-edelman-nictiz** [Nictiz/Nictiz-R4-zib2020#53]
Sometimes FHIR defines elements with a _required_ binding, whereas the zib prescribes a different ValueSet for the concept. For elements of type _code_, the CodeSpecification-extension is used (combined with a ConceptMap). However, for elements of type _CodeableConcept_, the expected way to handle this is to add a second `CodeableConcept.coding`.

Guidance needs to be developed on these topics:
* How to prescribe in the profile that the second ValueSet should be used?
* When is the addition of the second coding required/expected? Only when the mapping is not _equal_ or in all circumstances?

**Comment by ArdonToonstra** [comment]
I performed a quick analysis of our code-specification usage in STU3 for zib2017 profiles. We have only added it to datatypes such as code and coding and one time a string. 

Shall we wait for making the exact guidance until we encounter this for real? Or do you know a case? 

Preliminary thoughts for when we encounter a CodeadbleConcept element that have a _required_ binding, the usage of slice on `CodeableConcept.coding` seems to be the way to go. We should only require an additional coding if the mapping is not _equal_ or _equivalent_. If equal, it may be added too.

**Comment by pieter-edelman-nictiz** [comment]
Yes, we do have a case: the concepts ProblemStatus and VerificationStatus of zib Problem, which are mapped to `Condition.clinicalStatus` and `Condition.verificationStatus`.

On the topic of equal mappings; if the zib defines a _required_ binding, and not all elements map to the FHIR ValueSet, shouldn't the FHIR ValueSet be constrained?

**Comment by Ashwin-nictiz** [comment]
This is also in the case Encounter.class

**Comment by ArdonToonstra** [comment]
For the completeness here the required binding strength rules as applicable on a CodeableConcept:
> The following rules apply when required bindings are used with the CodeableConcept data type:
•	at least one Coding element SHALL be present
•	one of the Coding values SHALL be from the specified value set
•	text can be provided as well, and is always recommended, but is not an acceptable substitute for the required code

Proposal (for addition in paragraph 7.3 in profiling guidelines):

- Profiling a `CodeableConcept.coding` is required when the mapping of codes contain a code that cannot be mapped  _equal_. 
- Slice `CodeableConcept.coding` into one or two open slices discriminated by value/$this based on a required binding on the `CodeableConcept.coding` level.
-Mapping to the concepts of the functional description is applied both to the FHIR element and the `CodeableConcept.coding` slice. 
   - Possibly a slice that recreates the FHIR element, see note below.
        - Slicename =  copy of the FHIR element name for the parent CodeableConcept
        - Binding = copy FHIR binding of parent CodeableConcept – (with copy of the ConceptMap reference extension too?)
        - Cardinality = 1..1
   - A slice that holds the zib terminology.
        - Slice name = “The slice is named after the full English name of the value set according to the functional description, in camelCased notation.”
        - Binding = zib valueset and _required_ to make a valid slice definition
        - Cardinality = zib concept cardinality (most likely 0..1) 
        - Comment = guidance on when to use this coding.
**Note**: The slice that copies the FHIR valueset binding may not be necessary because FHIR dictates the required rules from the CodeadbleConcept level. Copying this information may however be clearer to the implementer. 

I made an example of both ways for `Condition.clinicalStatus` and `Condition.verificationStatus` as shown here in the image 
![image](https://user-images.githubusercontent.com/20518398/118642548-cc67af00-b7db-11eb-9db4-361cdf841d3a.png)

Made two examples: one valid and another invalid to test this out. Seems to work as expected.

**Comment by ArdonToonstra** [comment]
I think we should go for only creating a slice for our zib terminology. Don't copy FHIR. 

> On the topic of equal mappings; if the zib defines a required binding, and not all elements map to the FHIR ValueSet, shouldn't the FHIR ValueSet be constrained

That’s a good question… For `Condition.verificationStatus` that would mean we would constrain the FHIR valueset by removing the Entered-in-error code. 
I think we should constrain the FHIR ValueSet to only the codes that match with the zib. However, I would like to not do this as this might be a lot of additional work and touches the open-world modelling principle. What’s your thought on this?

@Ashwin-nictiz 
>This is also in the case Encounter.class
Encounter.class is of type Coding.

**Comment by pieter-edelman-nictiz** [comment]
Nice work @ArdonToonstra . I can see benefits and drawbacks of both approaches. Maybe we should put a textual hint in the slice documentation that the base FHIR binding still applies, and link to the relevant ConceptMap. In addition to that: shouldn't the ConceptMap be referenced using the extension as well?

I'm undecided on the requirement to include the zib code only when the code is not _equal_/_equivalent_. Wouldn't it be easier to always require the zib code when the ValueSet as a whole is incompatible, and not when it is compatible?

**Comment by ArdonToonstra** [comment]
Agreed with the textual hint in the slice documentation (I proposed on the comment of the coding) including the mention of an available ConceptMap. I am not sure if you want the ConceptMap extension here too? Because you will not map codes to this particular binding.  But I am not completely opposed to the idea. 

>I'm undecided on the requirement to include the zib code only when the code is not equal/equivalent. Wouldn't it be easier to always require the zib code when the ValueSet as a whole is incompatible, and not when it is compatible?

This might potentially give a lot of work to systems that do not natively support zib terminology. Some research data platforms are modelling EPD data into a data model based on FHIR and our zib profiles. If they want to exchange data based on our zib profiles, they will need to (manually) map their internal codes to both FHIR and zibs terminology while this might not be necessary for all codes. @LilianMinne can you comment on this to see if I am making a correct point here?

**Comment by pieter-edelman-nictiz** [comment]
> Agreed with the textual hint in the slice documentation (I proposed on the comment of the coding)

Oops, missed that one, sorry

**Comment by pieter-edelman-nictiz** [comment]
ConceptMaps are there to define the relationship between concepts, not specifically between a non-bound and a bound ValueSet. I think it makes a lot of sense to formally include it here, it tells the implementer how to go from the zib terminology to the FHIR terminology.

**Comment by pieter-edelman-nictiz** [comment]
> This might potentially give a lot of work to systems that do not natively support zib terminology. Some research data platforms are modelling EPD data into a data model based on FHIR and our zib profiles. If they want to exchange data based on our zib profiles, they will need to (manually) map their internal codes to both FHIR and zibs terminology while this might not be necessary for all codes.

Yeah, but you will always have to do this when the ValueSets are not fully equivalent, unless you can somehow guarantee that you will never send nor receive the codes that are not equivalent. So you need to create some complex logic for a fairly rare use case.

I also wonder why you need to map to both the zib and FHIR ValueSets. When you're doing zibs, you deal with zib ValueSets and use the given ConceptMap to go from there to the FHIR ValueSet.

**Comment by LilianMinne** [comment]
@ArdonToonstra @pieter-edelman-nictiz it's indeed true that some hospitals map their source data in the research data platform directly to FHIR and it's additional work for them to map to ZIB's as well. It's true that the mapping can be done to ZIB's first and then use ConceptMaps to get to FHIR, but this is not the way it's currently implemented. But as they need to change this to be able to qualify for MedMij, it doesn't really matter if they have to do this for all cases or only for cases where the mapping is not equivalent.

**Comment by ArdonToonstra** [comment]
Just spoke with Lilian and other colleagues working as data modellers and indeed concluded that it seems more logical to expect zib terminology for all concepts if a complete ValueSet cannot be mapped perfectly.

**Comment by pieter-edelman-nictiz** [comment]
I seems like we're reaching a consensus. To improve on @ArdonToonstra's earlier proposal

- A ConceptMap will be created to translate from the ValueSet of the functional model to the FHIR core ValueSet.
- Profiling `CodeableConcept.coding` is required when not all the codes in the ValueSet from the functional description can be mapped to the base FHIR ValueSet with an equivalence of _equivalent_ or _equal_. 
- `CodeableConcept.coding` will be sliced with a discriminator of value/$this, based on a required binding on the `CodeableConcept.coding` level.
- A slice will be created that binds the ValueSet of the functional description:
  - The slice is named after the full English name of the value set according to the functional description, in camelCased notation.
  - Binding strength = _required_ to make a valid slice definition.
  - The ConceptMap is documented using the permitted-value-conceptmap extension.
  - Cardinality = 1..1 (when the concept itself is included, bot the FHIR and function model codes are required).
  - Comment = A default string to describe the relation to the base binding. Proposal: "In addition to a coding from this ValueSet, the corresponding coding from the FHIR base ValueSet SHOULD be communicated. The ConceptMap <[canonical of ConceptMap]> can be used to relate these two ValueSets."
- Mapping to the concepts of the functional description is applied only to the FHIR element.

**Comment by ArdonToonstra** [comment]
Yes, I think we have a consensus. Nicely worded Pieter. I do have the next comments
> "In addition to a coding from this ValueSet, the corresponding coding from the FHIR base ValueSet SHOULD be communicated. The ConceptMap <[canonical of ConceptMap]> can be used to relate these two ValueSets."
If it concerns a required binding from the FHIR base value set it concerns a SHALL from the FHIR spec itself. We provide a SHOULD because we forse this construct for extensible bindings too?

> Mapping to the concepts of the functional description is applied only to the FHIR element. 
This includes the ConceptMap reference with the permitted-value-conceptmap extension too, right?

**Comment by pieter-edelman-nictiz** [comment]
> If it concerns a required binding from the FHIR base value set it concerns a SHALL from the FHIR spec itself. We provide a SHOULD because we forse this construct for extensible bindings too?

Hmm ... maybe I didn't think this through. I think it should be a "SHALL". Extensible doesn't mean you can do just whatever if there's also a alternative code from another code system.

> This includes the ConceptMap reference with the permitted-value-conceptmap extension too, right?

I meant the "root" of the concept, thus including the zib slice and including the permitted-value-conceptmap extension.

**Comment by ArdonToonstra** [comment]
Ok! We will go for SHALL. I think we can wrap this up in the Profiling Guidelines.

**Comment by pieter-edelman-nictiz** [comment]
Added this guidance to the profiling guidelines

**Comment by jduwel** [comment]
When profiling AllergyIntolerance following the guidelines (https://informatiestandaarden.nictiz.nl/wiki/FHIR:V1.0_FHIR_Profiling_Guidelines_R4#Metadata_for_elements) I noticed 2 inconsistencies:

* When slicing `coding`, the permitted-value-conceptmap extension is added to the *slice* ValueSet `.binding`, with `binding.description` set to: "_In addition to a coding from this ValueSet, the corresponding coding from the FHIR base ValueSet SHALL be communicated. The ConceptMap <[canonical of ConceptMap]> can be used to relate these two ValueSets._"
* When adding the `CodeSpecification extension, the permitted-value-conceptmap extension is added to the *FHIR* ValueSet `.binding`, with `binding.description` set to: "Use ConceptMap [ConceptMap.name] to translate terminology from the functional model to profile terminology in ValueSet [ValueSet.name]" (https://informatiestandaarden.nictiz.nl/wiki/FHIR:V1.0_FHIR_Profiling_Guidelines_R4#Metadata_for_elements)

I would argue that it is more consistent to: 
* Always add the permitted-value-conceptmap extension to the FHIR element
* Always add "_Use ConceptMap [ConceptMap.name] to translate terminology from the functional model to profile terminology in ValueSet [ValueSet.name]_" to the FHIR element `binding.description`.
* Add "_In addition to a coding from this ValueSet, the corresponding coding from the FHIR base ValueSet SHALL be communicated. The ConceptMap <[canonical of ConceptMap]> can be used to relate these two ValueSets._" to `comment` of the `.coding` slice.

The only situation wel `.coding` is sliced at the moment in the integration branch seems zib-Problem, and here the three bullets above are followed instead of the guidelines.

**Comment by jduwel** [comment]
I was not completely right when I said that zib-Problem already follows the proposal above. It seems `Condition.verificationStatus` does not refer to the ConcetMap using the permitted-value-conceptmap extension, only by `comment` on the slice.

**Comment by ArdonToonstra** [comment]
I think your proposal is an improvement, and I would think we should always add the permitted-value-conceptmap extension to the FHIR element.
We had a discussion in this issue whether to include the extension on both elements. I believe the consensus was to add them to both?

>ArdonToonstra commented on May 18
Agreed with the textual hint in the slice documentation (I proposed on the comment of the coding) including the mention of an available ConceptMap. I am not sure if you want the ConceptMap extension here too? Because you will not map codes to this particular binding. But I am not completely opposed to the idea.

> pieter-edelman-nictiz commented on May 18
ConceptMaps are there to define the relationship between concepts, not specifically between a non-bound and a bound ValueSet. I think it makes a lot of sense to formally include it here, it tells the implementer how to go from the zib terminology to the FHIR terminology.

**Comment by jduwel** [comment]
When reading the permitted-value-conceptmap definition (https://www.hl7.org/fhir/extension-11179-permitted-value-conceptmap.html) I notice a few things:

* It should be used together with https://www.hl7.org/fhir/extension-11179-permitted-value-valueset.html, but if we agree that permitted-value-valueset in our case _always_ is a zib valueset, we can get away with that.
> The source for the concept map is the value set the extension appears on. The target is the permitted-value-valueset.
* The extension should appear on the FHIR ValueSet, BUT this ValueSet should be the source, while the target is the zib ValueSet. At the moment we do this the other way around.
